### PR TITLE
Fix various bugs in BB8 : Closes #5571

### LINKED ISSUE #5571: Various bugs affecting BB8
Motivation
----------
Various bugs affecting BB8 were identified

### PR DESCRIPTION
**Pull request by cserf**
Fix various bugs in BB8 : Closes #5571

Add `allowed_scopes` option

Move to new heartbeat handler

Black formatting

### CODE REVIEW

File: bin/rucio-bb8 [hunk 1]
```diff
@@ -28,8 +29,11 @@ def get_parser():
     Returns the argparse parser.
     """
     parser = argparse.ArgumentParser(description='The BB8 daemon is responsible for rebalancing data between RSEs.')
-    parser.add_argument('rse', action='store', help='RSE to rebalance')
+    parser.add_argument('rse', action='store', help='RSE to rebalance. Can be either a RSE or RSE expression.')
```
Comment (mlassnig) [inline bin/rucio-bb8]: should be `--rse` ?

File: bin/rucio-bb8 [hunk 2]
```diff
@@ -28,8 +29,11 @@ def get_parser():
...
     parser.add_argument('bytes', action='store', type=int, help='Number of bytes to rebalance')
```
Comment (mlassnig) [inline bin/rucio-bb8]: should be `--bytes` ? also, not clear if this a goal without guarantees, or a minimum that will be used ?

File: lib/rucio/daemons/bb8/common.py [hunk 3]
```diff
@@ -112,28 +132,46 @@ def __dump_url(rse_id, logger=logging.log):
...
-        dump_dates = [dump_date.strftime('%d-%m-%Y')]
+        dump_dates = [dump_date.strftime("%d-%m-%Y")]
 
     # getting structure (template) of url location of a dump
-    url_template_str = config_get('bb8', 'dump_url_template', raise_exception=False, default='http://rucio-analytix.cern.ch:8080/LOCKS/GetFileFromHDFS?date=${date}&rse=${rse}')
+    url_template_str = config_get(
+        "bb8",
+        "dump_url_template",
+        raise_exception=False,
+        default="http://rucio-analytix.cern.ch:8080/LOCKS/GetFileFromHDFS?date=${date}&rse=${rse}",
```
Comment (mlassnig) [inline lib/rucio/daemons/bb8/common.py]: is this supposed to be the real URL? because then it's wrong. :-)

Comment (cserf) [reply lib/rucio/daemons/bb8/common.py]: :man_shrugging: That's the one that was used previously

File: bin/rucio-bb8 (hunk 1 above)
Comment (cserf) [reply bin/rucio-bb8]: Done

File: bin/rucio-bb8 (hunk 2 above)
Comment (cserf) [reply bin/rucio-bb8]: Done

File: lib/rucio/daemons/bb8/common.py (hunk 3 above)
Comment (cserf) [reply lib/rucio/daemons/bb8/common.py]: Remove default

File: lib/rucio/daemons/bb8/common.py [hunk 4]
```diff
@@ -221,100 +264,239 @@ def list_rebalance_rule_candidates(rse_id, mode=None, session=None):
...
+    query = (
+        session.query(
+            models.DatasetLock.scope,
+            models.DatasetLock.name,
+            models.ReplicationRule.id,
+            models.ReplicationRule.rse_expression,
+            models.ReplicationRule.subscription_id,
+            models.DataIdentifier.bytes,
+            models.DataIdentifier.length,
+            case(
```
Comment (rcarpa) [inline lib/rucio/daemons/bb8/common.py:420]: this new formatting makes me want to tear my eyes off

**Review by rcarpa — APPROVED** [review]

File: lib/rucio/daemons/bb8/common.py (hunk 4 above)
Comment (mlassnig) [reply lib/rucio/daemons/bb8/common.py:420]: I agree at least in this case.... usually I don't mind so much but this introduces way too much verticality. Much easier to understand what's going on if you don't have to scroll.

Comment (cserf) [reply lib/rucio/daemons/bb8/common.py:420]: That's much better on my 4:3 display :trollface:

Comment (bari12) [reply lib/rucio/daemons/bb8/common.py:420]: +1 I find that one also extremely unreadable 😢

**Pull request merged**